### ISSUE TITLE
Support ESM module

### ISSUE THREAD

**Issue by huan** [Chatie/tsconfig#16]
## Related links

- https://github.com/TypeStrong/ts-node/issues/1007
- https://github.com/TypeStrong/ts-node/issues/935
- [Pure ESM Package](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c)

**Comment by suntong** [comment]
> ## Howto
> * [Node Modules at War: Why CommonJS and ES Modules Can’t Get Along](https://redfin.engineering/node-modules-at-war-why-commonjs-and-es-modules-cant-get-along-9617135eeca1)
> * [How to Create a Hybrid NPM Module for ESM and CommonJS.](https://www.sensedeep.com/blog/posts/2021/how-to-create-single-source-npm-module.html)
> * [The ESM move #15](https://github.com/sindresorhus/meta/discussions/15)
> * [Publishing Node modules with TypeScript and ES modules](https://blog.logrocket.com/publishing-node-modules-typescript-es-modules/)

Thanks for providing all those informative links for us to read @huan, So

> Switching the default from CJS to ESM would be a big break in backwards compatibility. ...
> CJS can’t require() ESM because of top-level await

Would you explain just in high level what your strategy is please @huan? Especially, are you going to follow the recommendation to

> Provide a CJS version of your library

Thanks!

**Comment by suntong** [comment]
I meant, this is how I've been consuming wechaty in my plain JavaScript app:

```js
const {
  Wechaty,
  ScanStatus,
  log,
}               = require('wechaty')
```

And I presume lots of JavaScript developers are doing the same thing as well. 

> Would you explain just in high level what your strategy is please

At the very least, what we JavaScript developers should do, after such transition is finished? Hopefully _"Dual-ESM-CJS-module"_ means that we can still use our existing scripts without huge alternations. thx.

**Comment by suntong** [comment]
> # Migrating
> 
> There are two ways to move your packages to ESM:
> 
> 
> 1. Pure ESM  
       This has the benefit that it's easier to set up. You just add `"type": "module"` to your package.json, require Node.js 12, update docs & code examples, and do a major release.
>  1. Dual --- ESM with a build step that transpiles a CommonJS fallback  
This requires you to also set up a build step and add a `exports` field to your package.json. [Read more](https://nodejs.org/api/packages.html#packages_dual_commonjs_es_module_packages)  
_Note that there are _[_many downsides_](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c#gistcomment-3850849)_._
> 
> Personally, I plan to do `1` as I think it's better to rip off the bandaid and push the ecosystem forward.

Hmm..., maybe it is time to take plan #`1`...

**Comment by huan** [comment]
Yes, the ESM is the future and we should take it as soon as possible.

However, for maximum compatibility and a better developer experience, the Wechaty community strategy is to build a Dual-CJS-ESM package, which is what this issue is tracking about.

**Comment by suntong** [comment]
> you can use either `import` or `require` for our package.

👍👍👍

**Comment by suntong** [comment]
> you can use either `import` or `require` for our package

I tried, but one works and one doesn't. 
the two code is almost the same. Here are the key changes:

<img width="760" alt="image" src="https://user-images.githubusercontent.com/422244/175830077-0d0be95f-08ac-49eb-b2c4-f4d20adf107f.png">

and now it is working. Ref https://github.com/padlocal/wechaty-puppet-padlocal-demo/issues/16

So I'd conclude that the ESM module support is still broken, unfortunately.
